### ISSUE TITLE
teachMy - AI Chatbot by Oula Suliman

### ISSUE THREAD

**Issue by valiantone** [TechLabs-Berlin/project_proposals#18]
### Which track are you doing?
DL

### What's the problem you are trying to solve?
You have a material that's hard to understand? A difficult subject at school, a legal contract or a medical document? If only these text files can explain themselves, rephrase, answer questions and interact with the user so they can learn better, make decisions and save time!

### What's your idea for a solution to the given "problem"?
A personal AI learning assistant that allows users to upload any learning material as PDF and start interacting with it in a chatbot

### Which tracks do you think could be involved?
- DL/DS
- WD

### How do you imagine Deep Learning and/or Data Science could contribute to this project?
1. Developing a chatbot using a large language model (llm), fine-tuning and instructing the llm to produce desired educational outcomes and avoid certain behaviours. LLMs are not explicitly covered in the track, but they are today's biggest NLP application.
2. This might also require some testing and data analysis with the help of DS techies.

### How do you imagine Web Development could contribute to this project?
1. Building an interface where the user can upload files
2. Building a chatbot interface to interact with our AI assistant.
3. backend deployment of the AI model.
4. Other features might include: chat in quiz mode, previous chats, summary pages, learning progress statistics (graphs), "insert your idea"

### What are the most important challenges & bottlenecks of your project?
1. deployment: tackling GPU issues beyond prototyping on my own device. In such an app, there's always GPU power required to embed the user PDF and run an open source llm. Easy to deploy solutions like OpenAI would not be for free.
2. development: LLMs don't necessarily require fine-tuning to data as they're already able to generalise to such task. However, there are endless possibilities to what they can produce so their effectiveness and harm needs to be developed, measured and tested in creative ways

### What are the most important success factors of your project?
1. a user facing educational AI should be accurate, reliable, friendly and not cause harm or bias.
2. The design should be fun to bring those boring documents to life

### Share your resources here:
1. similar projects: plenty for search "LLM RAG" (RAG = Retrieval Augmented Generation)
- https://www.youtube.com/watch?v=LhnCsygAvzY&t=1615s&ab_channel=JamesBriggs
- https://medium.com/@indradumnabanerjee/building-rag-based-llm-application-with-custom-pdf-code-walkthrough-55cb38cb7ba6
- https://github.com/mananjain02/llm-custom-data2. 
2. LLM & RAG evaluation: 
- https://www.deepset.ai/blog/rag-evaluation-retrieval
- https://www.trulens.org/welcome/
3. LLM studio (does same thing on LARGE scale): 
- https://gpt.h2o.ai/

### Would you like to add an optional Pitch deck to support your idea in 6 to 8 slides?
- [Miro](https://miro.com/welcomeonboard/UTFoc0VtZHluYnV3TFNlQ2tWV1gwUmpjRVNUcEY4OVJJdHRadmpFZUM0dkFTUGtLYmxacVhPUlljZGhKZWliU3wzMDc0NDU3MzQ3NTkzNzg5NzI1fDI=?share_link_id=998181753473)

### Do you have any other remarks?
I think this idea has a straightforward MVP, and possibilities for improvements and added features

**Comment by valiantone** [comment]
Hey @OulaSu so first, great work on the proposal, it is well thought and covers a lot of the necessary descriptions. Just a few remarks:

1. Most of our projects will need a team of UX techies; without this the WD team usually does not have much direction and also the UI will lack purpose. For these reasons I think it would be great to include UX techies on this as well so your general application look and feel and functionality is better thought-out and more impressive. Please be open to including UX researchers to your project!

2. **Your task may be very challenging if you want to implement any fine-tuning of LLMs. This is slightly out of scope for even DL techies!** To minimize the risk I would recommend, first attempting the task of perhaps extracting content from any uploaded document and transform the learning information in to an interactive quiz in Step I. This is more feasible within the scope of a 10 week project MVP. If you are able to achieve this only then  attempt to create a fully conversational bot or AI agent with which you can have immersive dialog about content. **Also, since our curriculum does not provide resources for LLMs yet, we cannot guarantee you will be able to find the free resources for training and implementation. For these reasons, definitely recommend approaching this with a more achievable goal for the MVP!**

Just leave your comments on the points highlighted above. For most purposes your pitch is ready to move in to team formation stage. Just would like to have some insight in how you would tackle the above mentioned issues. We can them move your project to `Accepted` and start discussing final steps for team formation.

**Comment by OulaSu** [comment]
Hi @valiantone ! thank you so much for the well thought feedback!

1. UX: I expected this would come and I only left it out due to my lack of knowledge about UX workabout in general. I'd be happy to include UX techies for sure to support good colab between us and WD! 

2. LLM-finetuning is indeed out of the picture because it requires a training set how and it's not necessary. So despite llms not being explicitly covered, I think if some of the DL techies got interested during the NLP section, the llm would be just another model type. Short courses from https://www.deeplearning.ai/ would be sufficient to extend some up-to-date knowledge and they're free.
MVP - In my head it seemed easier to build a chatbot agent than a quiz builder so I'm not sure but there's a valid case for either way!  Would it be ok to keep this open for mutual decision when a team forms and we have more input?

**Comment by valiantone** [comment]
> MVP - In my head it seemed easier to build a chatbot agent than a quiz builder so I'm not sure but there's a valid case for either way! Would it be ok to keep this open for mutual decision when a team forms and we have more input?

Absolutely @OulaSu , let's leave it for a team decision once you have a few more DL techies added to the project. **Just to make it easier for me then would you be able to think of a simple back-up in case the team thinks a full chat agent would be too complex?**

One last question before we can start finalizing this project and discussing team formation. **Are there any tasks that you can imagine (list down a few) for the DS techies to perform?**

**Comment by OulaSu** [comment]
Yes! let's call it as the backup or option B MVP. At the end the goal is to interact with unstructured text data and both ideas would hit that.

the role of DS techies: for now I imagine we would do testing of the AI system to test for desired outcomes like accuracy, relevance, etc and we need to visualise, interpret, monitor and compare results across multiple prototype developments.

**Comment by valiantone** [comment]
ok then @OulaSu do you imagine 2 DS techies for this same? We can then have a team composition of 2 from each track?

**Comment by OulaSu** [comment]
Yes that's exactly what I thought be be a good aim: 2 of each track + me. so we'll be 3 DL techies since it's mainly a DL project

**Comment by valiantone** [comment]
All right, let's get started, would be nice to have 3DL techies, but let's put a minimum two out there for now. Go ahead and proceed with team formation!

### Team Formation Requirements

- [ ] 2-3 DL
- [ ] 2 DS
- [ ] 2 UX
- [ ] 1BE (WD)
- [ ] 1FE (WD)

**Comment by OulaSu** [comment]
thank you for your input!

**Comment by Fernanda-Portieri** [comment]
Hi @OulaSu, congrats for the proposal! I got very interested in your idea and I would like very much the opportunity to collaborate to the creation of this MVP.
@valiantone Zubin, I'm not entirely sure if I can pitch for a spot on the team, but I'm really keen on being part of it. If it works out, I'd be super excited to join the crew!

**Comment by OulaSu** [comment]
@Fernanda-Portieri thank you! yes you're very welcome to pitch in! I think we're encouraged to chose our own team from TechLab's side but Zubin can correct me ;) if you could just share what track you're in and which part you'd be interested in doing by joining! thoughts on the project itself are very welcome as well, here, on slack [bl-project_ideas] or the Miro board!

**Comment by Fernanda-Portieri** [comment]
Sorry @OulaSu, I forgot to mention, I am in the Data Science track. I just saw the board you created and I will study a bit more (alongside LLM). We'll catch up soon to discuss potential ways I can contribute.

**Comment by dusarp** [comment]
Hi @OulaSu, nice idea. I would be happy to join the team.

**Comment by OulaSu** [comment]
@dusarp sure thing! sorry your name is not visible, just let me know the name and track so I can update the project :)

**Comment by dusarp** [comment]
@OulaSu Arpad Dusa, DL techie. Thank you :)

**Comment by OulaSu** [comment]
welcome on board @dusarp ! with you, the DL team (3 people) is complete!

**Comment by mjkamegawa** [comment]
HI @OulaSu , I am in the UX track and I would love to be part of it as a UX designer. :D
